File: frontend/src/services/distance.js
const toRadians = (degrees) => degrees * (Math.PI / 180);

export const calculateDistance = (restLatitude, restLongitude, userLatitude, userLongitude) => {
    const R = 6371; // Radius of the Earth in kilometers
    const dLat = toRadians(restLatitude - userLatitude);
    const dLon = toRadians(restLongitude - userLongitude);
    const lat1 = toRadians(userLatitude);
    const lat2 = toRadians(restLatitude);

    const a = Math.sin(dLat / 2) * Math.sin(dLat / 2) +
              Math.sin(dLon / 2) * Math.sin(dLon / 2) * Math.cos(lat1) * Math.cos(lat2);
    const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
    const distance = R * c;

    return distance.toFixed(3);
};

File: backend/distance.py
import math


def to_radians(degrees):
    return degrees * (math.pi / 180)


def calculate_distance(rest_latitude, rest_longitude, user_latitude, user_longitude):
    R = 6371

    d_lat = to_radians(rest_latitude - user_latitude)
    d_lon = to_radians(rest_longitude - user_longitude)
    lat1 = to_radians(user_latitude)
    lat2 = to_radians(rest_latitude)

    a = math.sin(d_lat / 2) ** 2 + math.sin(d_lon / 2) ** 2 * math.cos(lat1) * math.cos(
        lat2
    )
    c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))
    distance = R * c

    return round(distance, 3)
